### ISSUE TITLE
ML Dog Bark

### ISSUE THREAD

**Issue by LievBarashkov1917** [LievBarashkov1917/Electronics-Codes#1]
/*
 * ========================================
 * ML DOG BARK DETECTOR - COMPLETE SYSTEM
 * ========================================
 * 
 * Hardware: MAX9814 Microphone + Arduino
 * Connection: MAX9814 OUT pin -> Arduino A0
 * 
 * This code has TWO MODES:
 * 1. TRAINING MODE: Collect samples for Edge Impulse
 * 2. INFERENCE MODE: Detect barks in real-time
 * 
 * Switch between modes by changing the MODE variable below
 */

// ========================================
// CONFIGURATION - CHANGE THIS
// ========================================
#define MODE 1  // Set to 1 for TRAINING, 2 for INFERENCE

// Hardware Configuration
const int MIC_PIN = A0;              // MAX9814 connected to analog pin A0
const int LED_PIN = 13;              // Built-in LED for visual feedback

// Audio Configuration
const int SAMPLE_RATE = 16000;       // 16kHz - standard for audio ML
const int BUFFER_SIZE = 16000;       // 1 second of audio (16000 samples)

// For INFERENCE MODE: Include your Edge Impulse library here
// Uncomment and replace with your project name after training:
// #include <your_project_name_inferencing.h>

// ========================================
// GLOBAL VARIABLES
// ========================================
int16_t audioBuffer[BUFFER_SIZE];    // Stores raw audio samples (0-1023)
int sampleIndex = 0;                 // Current position in buffer

// ========================================
// SETUP - RUNS ONCE AT STARTUP
// ========================================
void setup() {
  // Initialize serial communication at high speed
  // 115200 baud is fast enough to transfer audio data
  Serial.begin(115200);
  
  // Configure hardware pins
  pinMode(MIC_PIN, INPUT);           // Set A0 to read microphone data
  pinMode(LED_PIN, OUTPUT);          // Set LED pin as output
  digitalWrite(LED_PIN, LOW);        // Turn off LED initially
  
  // Wait for serial connection (helpful for debugging)
  while (!Serial) {
    delay(10);
  }
  
  // Print startup information based on mode
  Serial.println("========================================");
  Serial.println("   ML DOG BARK DETECTOR");
  Serial.println("========================================");
  
  if (MODE == 1) {
    // TRAINING MODE instructions
    Serial.println("MODE: TRAINING DATA COLLECTION");
    Serial.println("----------------------------------------");
    Serial.println("Instructions:");
    Serial.println("1. Open Serial Monitor");
    Serial.println("2. When you hear a dog bark:");
    Serial.println("   Type 'b' and press Enter");
    Serial.println("3. For background noise:");
    Serial.println("   Type 'n' and press Enter");
    Serial.println("4. Collect 20-30 samples of each");
    Serial.println("----------------------------------------");
    Serial.println("Ready to collect samples!");
  } 
  else if (MODE == 2) {
    // INFERENCE MODE instructions
    Serial.println("MODE: REAL-TIME BARK DETECTION");
    Serial.println("----------------------------------------");
    Serial.println("Listening for dog barks...");
    Serial.println("LED will light up when bark detected");
    Serial.println("----------------------------------------");
  }
  
  Serial.println();
  delay(1000);  // Give user time to read
}

// ========================================
// MAIN LOOP - RUNS CONTINUOUSLY
// ========================================
void loop() {
  if (MODE == 1) {
    // ========================================
    // TRAINING MODE: Collect labeled samples
    // ========================================
    trainingMode();
  } 
  else if (MODE == 2) {
    // ========================================
    // INFERENCE MODE: Detect barks in real-time
    // ========================================
    inferenceMode();
  }
}

// ========================================
// TRAINING MODE FUNCTION
// ========================================
void trainingMode() {
  // Wait for user to send a command via Serial Monitor
  if (Serial.available() > 0) {
    char command = Serial.read();    // Read the command character
    
    // Check if command is 'b' (bark) or 'n' (noise)
    if (command == 'b' || command == 'n') {
      // Determine the label based on command
      String label = (command == 'b') ? "bark" : "noise";
      
      // Visual feedback: blink LED to show recording started
      digitalWrite(LED_PIN, HIGH);
      Serial.print("Recording ");
      Serial.print(label);
      Serial.println(" sample...");
      
      // STEP 1: Collect audio sample
      collectAudioSample();
      
      // STEP 2: Send data to computer in Edge Impulse format
      sendTrainingData(label);
      
      // Turn off LED
      digitalWrite(LED_PIN, LOW);
      Serial.println("Sample recorded successfully!");
      Serial.println("----------------------------------------");
    }
    else {
      // Invalid command - remind user of correct commands
      Serial.println("Invalid command. Use 'b' for bark or 'n' for noise.");
    }
  }
}

// ========================================
// INFERENCE MODE FUNCTION
// ========================================
void inferenceMode() {
  /*
   * NOTE: This function requires the Edge Impulse library
   * You must first:
   * 1. Complete training on Edge Impulse
   * 2. Download and install the Arduino library
   * 3. Uncomment the #include line at the top
   * 4. Uncomment the code below
   */
  
  Serial.println("ERROR: Edge Impulse library not included!");
  Serial.println("Please:");
  Serial.println("1. Train your model on Edge Impulse");
  Serial.println("2. Download the Arduino library");
  Serial.println("3. Uncomment the #include line");
  Serial.println("4. Uncomment the inference code");
  delay(5000);
  
  /* UNCOMMENT THIS SECTION AFTER TRAINING:
  
  // STEP 1: Collect 1 second of audio
  Serial.println("Listening...");
  collectAudioSample();
  
  // STEP 2: Prepare data structure for Edge Impulse
  signal_t signal;
  signal.total_length = BUFFER_SIZE;
  signal.get_data = &getAudioData;
  
  // STEP 3: Run the neural network on our audio
  ei_impulse_result_t result = {0};
  EI_IMPULSE_ERROR res = run_classifier(&signal, &result, false);
  
  // STEP 4: Check for errors
  if (res != EI_IMPULSE_OK) {
    Serial.print("ERROR: Classification failed (");
    Serial.print(res);
    Serial.println(")");
    return;
  }
  
  // STEP 5: Print all predictions
  Serial.println("Predictions:");
  for (size_t i = 0; i < EI_CLASSIFIER_LABEL_COUNT; i++) {
    Serial.print("  ");
    Serial.print(result.classification[i].label);
    Serial.print(": ");
    Serial.print(result.classification[i].value * 100, 1);
    Serial.println("%");
  }
  
  // STEP 6: Check if bark detected with high confidence
  // Assuming first classification is "bark"
  float barkConfidence = result.classification[0].value;
  
  if (barkConfidence > 0.80) {  // 80% confidence threshold
    // BARK DETECTED!
    digitalWrite(LED_PIN, HIGH);
    Serial.println(">>> BARK DETECTED! <<<");
    Serial.print("Confidence: ");
    Serial.print(barkConfidence * 100, 1);
    Serial.println("%");
    delay(500);  // Keep LED on for 0.5 seconds
    digitalWrite(LED_PIN, LOW);
  }
  
  Serial.println("----------------------------------------");
  delay(100);  // Small delay before next check
  
  */
}

// ========================================
// FUNCTION: Collect Audio Sample
// ========================================
// This function records exactly 1 second of audio
// from the MAX9814 microphone at 16kHz sample rate
void collectAudioSample() {
  // Calculate microseconds between each sample
  // 1 second = 1,000,000 microseconds
  // 16,000 samples per second = 62.5 microseconds per sample
  unsigned long sampleInterval = 1000000 / SAMPLE_RATE;
  
  // Record BUFFER_SIZE samples (16,000 = 1 second)
  for (int i = 0; i < BUFFER_SIZE; i++) {
    // Mark the start time for this sample
    unsigned long startTime = micros();
    
    // Read microphone value (0-1023 from Arduino's 10-bit ADC)
    audioBuffer[i] = analogRead(MIC_PIN);
    
    // Wait until it's time for the next sample
    // This loop ensures consistent timing (exactly 16kHz)
    while (micros() - startTime < sampleInterval) {
      // Do nothing - just wait for precise timing
    }
  }
}

// ========================================
// FUNCTION: Send Training Data
// ========================================
// Formats and sends audio data to Edge Impulse
// via Serial in a specific format they can parse
void sendTrainingData(String label) {
  // Print label header
  Serial.print("LABEL:");
  Serial.println(label);
  
  // Print data header
  Serial.println("DATA:");
  
  // Send all 16,000 audio samples
  // Each sample on a new line
  for (int i = 0; i < BUFFER_SIZE; i++) {
    Serial.println(audioBuffer[i]);
  }
  
  // Print end marker
  Serial.println("END");
}

// ========================================
// FUNCTION: Get Audio Data (Callback)
// ========================================
// This is a callback function for Edge Impulse
// The neural network calls this to get audio data in chunks
// 
// Parameters:
//   offset: Starting position in buffer
//   length: How many samples to return
//   out_ptr: Where to put the data
int getAudioData(size_t offset, size_t length, float *out_ptr) {
  // Copy the requested portion of audio buffer
  for (size_t i = 0; i < length; i++) {
    // Convert int16_t to float and store in output array
    out_ptr[i] = (float)audioBuffer[offset + i];
  }
  return 0;  // Return 0 = success
}

/*
 * ========================================
 * HOW THIS ENTIRE SYSTEM WORKS
 * ========================================
 * 
 * PHASE 1: TRAINING (MODE = 1)
 * -----------------------------
 * 1. Arduino waits for your command ('b' or 'n')
 * 2. When command received:
 *    - Turns on LED (recording indicator)
 *    - Collects 1 second of audio (16,000 samples)
 *    - Sends data via Serial in Edge Impulse format
 * 3. You copy/paste this data into Edge Impulse
 * 4. Repeat 20-30 times for each class (bark/noise)
 * 5. Edge Impulse trains a neural network on this data
 * 
 * 
 * PHASE 2: INFERENCE (MODE = 2)
 * -------------------------------
 * 1. Arduino continuously listens (every ~1 second)
 * 2. For each sample:
 *    - Collects 1 second of audio
 *    - Edge Impulse library extracts MFCC features
 *      (MFCC = audio fingerprint that ML models understand)
 *    - Neural network analyzes features
 *    - Returns probability for each class:
 *      Example: bark=0.92, noise=0.08 (92% sure it's a bark)
 * 3. If bark probability > 80%:
 *    - Lights up LED
 *    - Prints "BARK DETECTED!"
 *    - You can add your own actions here
 * 
 * 
 * WHY THIS WORKS BETTER THAN SIMPLE THRESHOLD:
 * ---------------------------------------------
 * Simple approach: if (volume > threshold) -> bark
 * Problems: 
 *   - False alarms from loud noises
 *   - Misses quiet barks
 *   - Can't distinguish bark from other sounds
 * 
 * ML approach: Learns actual bark patterns
 * Benefits:
 *   - Recognizes bark characteristics (frequency, rhythm)
 *   - Works with different dogs and volumes
 *   - Ignores non-bark sounds even if loud
 *   - 85-95% accuracy with good training data
 * 
 * 
 * TECHNICAL DETAILS:
 * ------------------
 * Sample Rate: 16kHz (16,000 samples per second)
 *   - Why? Dog barks are 500-2000 Hz
 *   - Nyquist: Need 2x highest frequency = 4kHz minimum
 *   - 16kHz is standard for speech/audio ML
 *   - Captures up to 8kHz (more than enough)
 * 
 * Buffer Size: 16,000 samples = 1 second
 *   - Enough to capture full bark (typically 0.1-0.4 seconds)
 *   - Not too long (uses ~32KB RAM)
 *   - Standard window size for audio classification
 * 
 * MFCC Features: 
 *   - Mel-Frequency Cepstral Coefficients
 *   - Converts time-domain audio to frequency representation
 *   - Similar to how human ears perceive sound
 *   - Compact representation (40 features from 16,000 samples)
 *   - Neural networks work much better with MFCC than raw audio
 * 
 * Neural Network:
 *   - Trained on Edge Impulse cloud
 *   - Deployed as optimized C++ library
 *   - Runs locally on Arduino (no internet needed)
 *   - Typical size: 10-50KB of flash memory
 *   - Inference time: 50-200ms per sample
 * 
 * 
 * MEMORY REQUIREMENTS:
 * --------------------
 * - audioBuffer: 32KB (16,000 samples × 2 bytes)
 * - Neural network: 10-50KB (depends on model)
 * - Total: ~50-100KB RAM needed
 * 
 * Compatible boards:
 *   ✓ Arduino Nano 33 BLE (256KB RAM)
 *   ✓ ESP32 (520KB RAM)
 *   ✓ Arduino Portenta (8MB RAM)
 *   ✗ Arduino Uno (2KB RAM - NOT ENOUGH)
 *   ✗ Arduino Nano (2KB RAM - NOT ENOUGH)
 * 
 * 
 * TROUBLESHOOTING:
 * ----------------
 * "Out of memory" error:
 *   - Reduce BUFFER_SIZE to 8000 (0.5 seconds)
 *   - Use int8 quantization in Edge Impulse
 * 
 * Low accuracy:
 *   - Collect more training samples (50+ per class)
 *   - Add more variety (different dogs, distances, environments)
 *   - Balance dataset (equal bark/noise samples)
 * 
 * False positives (detecting barks when there aren't any):
 *   - Increase confidence threshold (0.80 -> 0.90)
 *   - Add more noise samples to training data
 * 
 * Missing real barks:
 *   - Lower confidence threshold (0.80 -> 0.70)
 *   - Add more bark samples from different dogs
 *   - Check if microphone gain is too low
 * 
 */
